File: spark-join-parity/code/JoinBench.scala
import org.apache.spark.sql.SparkSession

case class Event(id: Long, ts: Long, amount: Double, category: String)
case class Account(id: Long, region: String, tier: Int, signup_epoch: Long)

// Overridable via env vars so the same script can run with a shorter
// schedule as one half of a concurrent pair (see run_concurrent.sh),
// without duplicating this file.
val RUNS = sys.env.getOrElse("BENCH_RUNS", "50").toInt
val WARMUP = sys.env.getOrElse("BENCH_WARMUP", "20").toInt

import spark.implicits._

val events = spark.read.parquet("/tmp/bench/data/events").as[Event]
val accounts = spark.read.parquet("/tmp/bench/data/accounts").as[Account]

for (i <- 0 until (WARMUP + RUNS)) {
  val tag = if (i < WARMUP) "warmup" else s"run_${i - WARMUP + 1}"
  sc.setJobGroup(tag, s"scala dataset joinWith - $tag")
  val joined = events.joinWith(accounts, events("id") === accounts("id"), "inner")
  val total = joined.count()
  println(s"[$tag] joined_count=$total")
}

sc.setJobGroup("", "")
System.exit(0)

File: spark-join-parity/code/join_bench.py
from pyspark.sql import SparkSession

RUNS = 10
WARMUP = 2

spark = (
    SparkSession.builder
    .master("local[4]")
    .appName("pyspark-dataframe-join-bench")
    .config("spark.driver.host", "127.0.0.1")
    .config("spark.driver.bindAddress", "127.0.0.1")
    .config("spark.driver.memory", "2g")
    .config("spark.sql.shuffle.partitions", "8")
    .config("spark.sql.autoBroadcastJoinThreshold", "-1")
    .config("spark.sql.adaptive.enabled", "false")
    .config("spark.eventLog.enabled", "true")
    .config("spark.eventLog.dir", "file:///tmp/bench/events/pyspark")
    .getOrCreate()
)
spark.sparkContext.setLogLevel("WARN")
sc = spark.sparkContext

events = spark.read.parquet("/tmp/bench/data/events")
accounts = spark.read.parquet("/tmp/bench/data/accounts")

for i in range(WARMUP + RUNS):
    tag = "warmup" if i < WARMUP else f"run_{i - WARMUP + 1}"
    sc.setJobGroup(tag, f"pyspark dataframe join - {tag}")
    joined = events.join(accounts, on="id", how="inner")
    total = joined.count()
    print(f"[{tag}] joined_count={total}")

sc.setJobGroup("", "")
spark.stop()
